### ISSUE TITLE
Investigate: Apply default profiles early during initialisation

### ISSUE THREAD

**Issue by glts** [technomancy/leiningen#2589]
During Leiningen [core initialisation](https://github.com/technomancy/leiningen/blob/2.9.1/leiningen-core/src/leiningen/core/project.clj#L974) there is a vicious circle in which the initial setup requires repository connection settings to be available, but these settings are available only after setup is complete.

In Leiningen 2.8.2, a **new configuration file [`repository-overrides.clj`](https://github.com/technomancy/leiningen/issues/2453)** was introduced to address this. This is a simple and low-cost solution for this problem. But it is in my opinion rather workaround-like, an afterthought. One problematic aspect of it is that it does not actually ‘override’: it is merged into the project map at the very beginning, but can itself be overridden later by settings defined in profiles.

Most users’ expectation in the past years has been that they should be able to **specify such settings in their user profile**. There have been many such reports, eg #1712 #1722 #2057 #2058 #2059 #2111 #2371.

It might be worth investigating if the whole problem can be solved within the existing profiles design, and de-emphasise and ultimately render obsolete the (undocumented) `repository-overrides.clj` mechanism. Perhaps the whole string of issues can be addressed simply by reading and merging default profiles early on during initialisation.

I’m interested in feedback (don’t know if at all viable or desirable? @ghaskins @technomancy) and might try to prototype something in a few days.

**Comment by ghaskins** [comment]
> It might be worth investigating if the whole problem can be solved within the existing profiles design, and de-emphasise and ultimately render obsolete the (undocumented) repository-overrides.clj mechanism. 

@glts id be in favor of this

**Comment by glts** [comment]
Pull request has been merged, thank you for the review.
